File: 1705-maximum-number-of-eaten-apples/1705-maximum-number-of-eaten-apples.cpp
class Solution {
public:
    int eatenApples(vector<int>& apples, vector<int>& days) {
        
        int i=0,n=apples.size(),cnt=0;
        priority_queue<pair<int,int>,vector<pair<int,int>>,greater<pair<int,int>>>pq;
        
        while(i<n || !pq.empty()){
            if(i<n && apples[i]!=0)
            pq.push({i+days[i],apples[i]});
        
            while(!pq.empty() && pq.top().first<i+1)pq.pop();
            
            if(!pq.empty()){
                auto [first,sec]=pq.top();pq.pop();
                cnt++;
                if(--sec>0)pq.push({first,sec});
                }
            i++;
            }
        
        return cnt;
        
    }
};

File: 1705-maximum-number-of-eaten-apples/1705-maximum-number-of-eaten-apples.java
class Solution {
    public int eatenApples(int[] apples, int[] days) {
        
        int cnt=0,i=0,n=apples.length;
        PriorityQueue<int[]>pq=new PriorityQueue<>((a,b)->a[0]-b[0]);
        
        while(i<n || !pq.isEmpty()){
            if(i<n && apples[i]!=0)pq.add(new int[]{i+days[i],apples[i]});
            
            while(!pq.isEmpty() && pq.peek()[0]<i+1)pq.poll();
            
            if(!pq.isEmpty()){
                if(pq.peek()[1]>1)pq.peek()[1]--;
                else pq.poll();
                cnt++;
            }
            i++;
        }
        return cnt;
    }
}